keel-core/src/main/kotlin/com/netflix/spinnaker/keel/serialization/mixins/ApiVersionMixin.kt

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2020 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.netflix.spinnaker.keel.serialization.mixins
+
+import com.fasterxml.jackson.annotation.JsonCreator
+import com.fasterxml.jackson.annotation.JsonIgnore
+import com.fasterxml.jackson.databind.annotation.JsonSerialize
+import com.fasterxml.jackson.databind.ser.std.ToStringSerializer
+import com.netflix.spinnaker.keel.api.ApiVersion
+
+@MixinFor(ApiVersion::class)
+@JsonSerialize(using = ToStringSerializer::class)
+abstract class ApiVersionMixin(
+  val group: String,
+  val version: String
+) {
+
+  @JsonCreator
+  constructor(value: String) :
+    this(value.substringBefore("/"), value.substringAfter("/"))
```

Comment:
This confuses me. We're constructing the mixin here, but this looks like constructing the actual type. What's going on?

Reply:
I have no idea how the internals work, but Jackson is somehow capable of defining constructors on the mixin and applying it to the actual type, like you're guessing. Not sure how it actually gets the logic of this constructor applied to the type, but it does work. 🤷‍♂